### DISCUSSION TITLE
Handling Android 15's edge-to-edge enforcement on React Native

### DISCUSSION THREAD

**Discussion by alanleedev** [react-native-community/discussions-and-proposals#827] [Announcements]
**tl;dr: Android 15 will now enforce edge-to-edge when you opt-in to targetSdk 35. For properly handling such scenario, we are recommending community library `react-native-safe-area-context`.**

## What is changing with Android 15?

Android 15 has a very noticeable change with how the app content is rendered on screen. When the app's `targetSdk` is set to `35`, the OS will enforce edge-to-edge when running on Android 15. This means that the content area will be expanded to render using the full screen and even under the status bar and the navigation bar area. Apps are expected to make UI adjustments to make sure important content does not overlap with system bars even if the app had not previously supported edge-to-edge. 

See Android documentation for more details;
- https://developer.android.com/about/versions/15/behavior-changes-15#edge-to-edge
- https://developer.android.com/develop/ui/views/layout/edge-to-edge

Note: Some props no longer work in this case as Android is deprecating APIs related to forced edge-to-edge. For example, system bars will be transparent and you won't be able to set the color of the status bar.

## Regarding the `<SafeAreaView>` component

[SafeAreaView](https://reactnative.dev/docs/next/safeareaview) component was added for iOS to support the feature of the same name since iOS 11. This allows content to be expanded to fill the available screen area without overlapping with system UI like the status bar or being cut-off due to rounded edges of the device. Android also has been supporting edge-to-edge in recent versions of the OS but there was no official support for this in the React Native core.

While `SafeAreaView` has worked in some cases, it is very limited in its usage due to it not offering any options to control its behavior. It only offers automatic padding to prevent content overlap or cut-off. Also did I mention it is iOS only?

Due to these reasons, we are moving away from using `SafeAreaView` and will be deprecating it in the future.

### react-native-safe-area-context
[react-native-safe-area-context](https://github.com/th3rdwave/react-native-safe-area-context) has been a popular community library with feature rich capabilities. We are recommending it to handle (forced) edge-to-edge on Android 15. It can be used to support general edge-to-edge in other Android versions and iOS as well. It provides SafeAreaView replacement with more options and additional hooks to get system inset values which offer more flexibility in its usage. 

Also code examples in [reactnative.dev](https://reactnative.dev/) website have been updated to use `react-native-safe-area-context` for version 0.76. Special thanks to [@Simek](https://github.com/Simek) for updating all the code examples.

###  react-native-edge-to-edge
Edge-to-edge isn't just about displaying content in a safe area. There are other gaps to fill if you want to fully embrace edge-to-edge. [react-native-edge-to-edge](https://github.com/zoontek/react-native-edge-to-edge) is a new library backed by Expo that has recently released version 1. It is trying to tackle gaps in general edge-to-edge support and also provides some recommendations in other areas where you may need to consider.

**Comment by krizzu** [comment]
I think it's also worth noting that there's an opt-out mechanism (as for now, it is now know when it'll be removed) as [this article describes it](https://www.androidauthority.com/android-15-edge-to-edge-opt-out-3467646/). This gives more time to adapt an app to edge-to-edge.

**Comment by vytautasvargonas** [comment]
just so you know `react-native-safe-area-context` returns 0 as bottom inset for Android 15. and enabling edge to edge / using `react-native-edge-to-edge` / setting android navigation bar as transparent breaks the Android keyboard avoiding of `windowSoftInputMode="adjustResize"`

**Reply by kirillzyusko** [reply]
@vytautasvargonas you can use `KeyboardAvoidingView` on Android to restore keyboard avoiding. But a slightly better approach would be the usage of [react-native-keyboard-controller](https://github.com/kirillzyusko/react-native-keyboard-controller) - it'll handle keyboard with animated transitions (which will significantly improve UX of the app).

**Reply by vytautasvargonas** [reply]
curious if anyone made it work with edge to edge enabled?

**Reply by kirillzyusko** [reply]
> curious if anyone made it work with edge to edge enabled?

@vytautasvargonas What's exactly? `KeyboardAvoidingView` or `react-native-keyboard-controller`? `react-native-keyboard-controller` works in edge-to-edge mode only (but optionally you can apply paddings to avoid overlap with system elements).

**Reply by zoontek** [reply]
> using react-native-edge-to-edge / setting android navigation bar as transparent breaks the Android keyboard avoiding of windowSoftInputMode="adjustResize"

That's indeed the Android behavior, it's explained in the README. Using `KeyboardAvoidingView` with `behavior="padding"` or `react-native-keyboard-controller` works.

**Reply by rmargolinross** [reply]
The 0 returned from react-native-safe-area-context is a huge issue currently and it seems there is no good solution for it yet. Has anyone found a solution that I do not know about?

**Reply by gavrichards** [reply]
Would also like to know if anyone has a workaround for this significant problem? We're positioning an element of our UI using the bottom inset, but as it returns 0 in Android 15 when it shouldn't, the element is positioned off the bottom of the screen.

**Comment by robrechtme** [comment]
Also important to mention is that the [`contentInsetAdjustmentBehavior`](https://reactnative.dev/docs/scrollview#contentinsetadjustmentbehavior-ios) prop on `ScrollView` also is iOS only, so you cannot rely on this to respect the safe area inside scroll views on Android.

**Comment by kazuya-sakamoto** [comment]
Currently, when running on Android 15 with the app’s targetSdk set to 35, adjustResize gets disabled. As a result, I decided to use [react-native-keyboard-controller](https://github.com/kirillzyusko/react-native-keyboard-controller). To enable react-native-keyboard-controller, you need to import { KeyboardProvider } from 'react-native-keyboard-controller' in your root component. Doing this allows useKeyboardHeight to properly manage the keyboard’s height.

I found the [“Animating Views in Sync with Keyboard Height”](https://docs.expo.dev/guides/keyboard-handling/#animating-views-in-sync-with-keyboard-height) article in the Expo documentation extremely helpful. Thanks to that guide, I was able to properly handle the keyboard overlap issue in my project. If anyone else is dealing with similar problems, I highly recommend checking it out.

**Reply by franos1910** [reply]
But it relies on reanimated, and reanimated is buggy when you want to use native Modal on screen with useAnimatedStyles..

**Reply by kirillzyusko** [reply]
@franos1910 the keyboard-controller API has corresponding hooks that are based on plain Animated API (which is provided by react-native itself).

If you are using pre-built components, then yes, some of them are based on reanimated, because default Animated API has its own fundamental limitations (such as inability to animate height and other properties)...

**Comment by DewantiV** [comment]
I'm thinking twice about implementing edge-to-edge for now

**Comment by a-klotz-p8** [comment]
Not sure if this is the right place to ask this. But I am developing an react native 0.78 app with a targetSdk of 35. I am using react-native-safe-area-context to handle the edge-to-edge. Now when I uploaded my bundle to Google Play Console I got the warning that the edge-to-edge might not work for some users and that I need to use enableEdgeToEdge() for Kotlin or EdgeToEdge.enable() for Java. 

Do I really need to add anything if I have safe area context already in use and tested it on a Android 15 device or can I just ignore this warning?
Somewhere I read that I might need to add WindowCompat.setDecorFitsSystemWindows(window, false) to my onCreate in the MainActivity.kt?

**Reply by zoontek** [reply]
You need react-native-edge-to-edge to enable it on Android < 15

**Reply by a-klotz-p8** [reply]
That makes sense, thanks. But if I only want to support it on android > 15, I am fine with my current setup?

**Reply by zoontek** [reply]
@a-klotz-p8 In this case, yes (But you'll keep the warnings on the Play Store console).

**Comment by getsettalk** [comment]
oh have anyone find proper solution for keyboard handling , because of react-native-keyboard-controller making extra work for handling each screen keyboard

**Reply by kirillzyusko** [reply]
@getsettalk if you want to have the same keyboard handling as before, then you can simply wrap your whole app inside `KeyboardAvoidingView` with `padding` mode.

But effectively keyboard-controller tries to make keyboard behavior identical on both platforms, so if you already have proper keyboard handling on iOS, then it should be only a matter of changing imports and replace the usage of other components (KeyboardAvoidingView, KeyboardAwareScrollView) with corresponding components from `react-native-keyboard-controller` 🤷‍♂️

**Comment by sarthakmetazone** [comment]
How we are handling the native screens, like for payment or chatbot?

**Comment by Prajwaltechversant** [comment]
I tried disabling edge-to-edge by setting edgeToEdgeEnabled={false}, but it is not working as expected in my project. I am using React Native Navigation, and the screen still renders in edge-to-edge mode.

**Reply by TheGipsyDanger** [reply]
Where did you make this configuration?

**Reply by liamjones** [reply]
gradle.properties if you're on a new enough version

**Reply by Prajwaltechversant** [reply]
> gradle.properties if you're on a new enough version

yeah i have added that in gradle.properties

**Reply by Prajwaltechversant** [reply]
> Where did you make this configuration?

gradle.properties

**Comment by tmdreepu** [comment]
Due to edge-to-edge enforcement introduced in Android 15, setting the status bar as translucent is deprecated in API level 35 and setting it will have no effect. You can read more about our [edge-to-edge recommendations here](https://github.com/react-native-community/discussions-and-proposals/discussions/827).  ,    i'm working api level 34, now fater update lavel , status bar  no effect , now what should I do ?

**Comment by JosDevv** [comment]
I have an application made with react native without expo and my solution was to add <item name="android:windowOptOutEdgeToEdgeEnforcement">true</item> in android\app\src\main\res\values\styles.xml

**Reply by farazreact** [reply]
> I have an application made with react native without expo and my solution was to add true in android\app\src\main\res\values\styles.xml

can you clearify what you added ?

**Reply by AbhinayMe** [reply]
Seems he have added the below line to the style 'AppTheme'

`<item name="android:windowOptOutEdgeToEdgeEnforcement">true</item>`

**Reply by zoontek** [reply]
@AbhinayMe This is a temporary solution that only exists on Android 15, not Android 16.

**Reply by AbhinayMe** [reply]
Totally agree! I was just trying to wrap up what he was saying. 😅
